specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+          ``distros`` namespace.
+        * lots of utilities, with low coherence between them.
+
+    - cloudbaseinit has a ``osutils`` location. There is a ``BaseOSUtils``
+      base class, with a WindowsUtils implementation.
+
+       Problems:
+
+           * it's a pure utilities class, leading to low coherence
+             between functions.
+           * it is not a namespace of OS specific functionality.
+             For this, there is also ``utils.windows``.
+
+As seen, both projects lack a namespaced location for all the OS related code.
+
+The following architecture proposal tries to solve this issue, by having one
+namespace for both general utilies related to a distro, as well as distro
+specific code, which doesn't have a counterpart on other distros.
+
+It can have the following advantages:
+
+    * one common place for all distro interaction, with standardized
+      API for each subnamespace and increased coherence.
+
+    * avoids leaky abstractions. Distro specific code goes into ``distros``
+      namespace.
+
+    * eases testability, it is easy to provide a mock with autospeccing
+      for a namespace class, such as Route.
+
+    * Pythonic API, easy to understand and use.
+
+
+The distros location is proposed, with the following structure and attributes:
+
+    - The base class for a distro is found in ``distros.base``.
+
+    - There are sub-namespaces for specific interaction with the OS,
+      such as network, users.
+
+    - More namespaces can be added, if we identify a group of interactions that can
+      be categorized in a namespace.
+
+    - There is a ``general`` namespace, which contains general utilities that can't be moved
+      in another namespace.
+
+    - Each subnamespace has its own abstract base class, which must be implemented
+      by different distros. Code reuse between distros is recommended.
+
+    - Each subnamespace exposes a way to obtain an instance of that namespace for
+      the underlying distro. This can be exposed in namespace.factory.
+      Thus, the following are equivalent:
+
+        >>> from cloudinit import distro
+        >>> network = distro.get_distro().network
+
+        >>> from cloudinit.distro.network import factory
+        >>> network = factory.get_network()
+
+    - Each subnamespace can expose additional behaviour that might not exist in
+      the base class, if that behaviour does not make sense or if there is no
+      equivalent on other platforms. But don't expose leaky abstraction, this
+      specific code must be written in an abstract way, so that possible alternatives
+      can be found for other distros in the mean time. This means that no ctypes
+      interaction with the Windows should be exposed,
+      but an object with a guaranteed interface.
+
+
+      cloudinit/distros/__init__.py
+                        base.py
+                        factory.py
+
+                        network/
```

Comment:
I'm actually agreeing with Daniel, too. My first intention was to separate the concerns from inside the distro into more useful and cohesive subnamespaces, but the side effect is an increased maintenance for adding custom distros, which didn't crossed my mind initially. I'll update the spec then.